File: packages/disable_screenshot/android/src/main/java/com/appinio/disable_screenshot/DisableScreenshotPlugin.java
package com.appinio.disable_screenshot;

import android.os.Build;
import android.os.Environment;
import android.os.FileObserver;
import android.os.Handler;
import android.os.Looper;
import android.view.WindowManager;
import android.webkit.MimeTypeMap;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import java.io.File;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Timer;
import java.util.TimerTask;

import io.flutter.embedding.engine.plugins.FlutterPlugin;
import io.flutter.embedding.engine.plugins.activity.ActivityAware;
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding;
import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;
import io.flutter.plugin.common.MethodChannel.MethodCallHandler;
import io.flutter.plugin.common.MethodChannel.Result;
import io.flutter.plugin.common.EventChannel;

/** DisableScreenshotPlugin */
public class DisableScreenshotPlugin implements FlutterPlugin, MethodCallHandler, EventChannel.StreamHandler, ActivityAware {
  private MethodChannel channel;
  private FileObserver fileObserver;
  private Timer timeout = new Timer();
  private final Map<String,FileObserver> watchModifier = new HashMap<>();
  private ActivityPluginBinding activityPluginBinding;
  private Handler handler;
  private boolean screenRecording = false;
  private long tempSize = 0;
  private EventChannel.EventSink eventSink = null;

  @Override
  public void onAttachedToEngine(@NonNull FlutterPluginBinding flutterPluginBinding) {
    channel = new MethodChannel(flutterPluginBinding.getBinaryMessenger(), "com.appinio.screenshot/disableScreenshots");
    EventChannel eventChannel = new EventChannel(flutterPluginBinding.getBinaryMessenger(), "com.appinio.screenshot/observer");
    channel.setMethodCallHandler(this);
    eventChannel.setStreamHandler(this);
  }

  @Override
  public void onListen(Object arguments, EventChannel.EventSink events) {
    this.eventSink = events;
    handler = new Handler(Looper.getMainLooper());
    updateScreenRecordStatus();
    if(Build.VERSION.SDK_INT >=29){
      final List<File> files = new ArrayList<>();
      final List<String> paths = new ArrayList<>();
      for (Path path : Path.values()) {
        files.add(new File(path.getPath()));
        paths.add(path.getPath());
      }
      fileObserver = new FileObserver(files) {
        @Override
        public void onEvent(int event, final String filename) {
          if (event == FileObserver.CREATE) {
            handler.post(() -> {
              for(String fullPath : paths){
                File file = new File(fullPath + filename);
                if(file.exists()){
                  if(getMimeType(file.getPath()).contains("image")){
                    eventSink.success("Got Screenshot");
                  }
                }
              }
            });
          }
        }
      };
      fileObserver.startWatching();
    }else{
      for (final Path path : Path.values()) {
        fileObserver = new FileObserver(path.getPath()) {
          @Override
          public void onEvent(int event, final String filename) {
            updateScreenRecordStatus();
            File file = new File(path.getPath() + filename);
            if (event == FileObserver.CREATE) {
              handler.post(() -> {
                if(file.exists()){
                  if(getMimeType(file.getPath()).contains("image")){
                    eventSink.success("Got Screenshot");
                  }
                }
              });
            }
          }
        };
        fileObserver.startWatching();
      }
    }
  }

  @Override
  public void onCancel(Object arguments) {
    if(fileObserver!=null) fileObserver.stopWatching();
    for (Map.Entry<String, FileObserver> stringObjectEntry : watchModifier.entrySet()) {
      stringObjectEntry.getValue().stopWatching();
    }
    watchModifier.clear();
    eventSink = null;
  }

  @Override
  public void onMethodCall(@NonNull MethodCall call, @NonNull Result result) {
    switch(call.method){
      case "disableScreenshots":
        if((boolean)call.arguments){
          activityPluginBinding.getActivity().getWindow().addFlags(WindowManager.LayoutParams.FLAG_SECURE);
        }else{
          activityPluginBinding.getActivity().getWindow().clearFlags(WindowManager.LayoutParams.FLAG_SECURE);
        }
        break;
      default:
    }
  }

  private void updateScreenRecordStatus(){
    List<String> paths = new ArrayList<>();
    for(Path path : Path.values()){
      paths.add(path.getPath());
    }
    for (int i = 0; i < Path.values().length;i++){
      String fullPath = paths.get(i);
      File newFile = getLastModified(fullPath);
      if(newFile!=null && getMimeType(newFile.getPath()).toLowerCase().contains("video") && !watchModifier.containsKey(newFile.getPath())){
        watchModifier.put(newFile.getPath(), new FileObserver(newFile) {
          @Override
          public void onEvent(int event, @Nullable String path) {
            long curSize = newFile.length();
            if(curSize > tempSize){
              if(timeout!=null){
                try {
                  timeout.cancel();
                  timeout = null;
                }catch (Exception ignored){}
              }
              setScreenRecordStatus(event == FileObserver.MODIFY);
              tempSize = newFile.length();
            }
            if(timeout == null){
              timeout = new Timer();
              timeout.schedule(new TimerTask() {
                @Override
                public void run() {
                  if(watchModifier.containsKey(newFile.getPath())){
                    setScreenRecordStatus(curSize != tempSize);
                  }
                }
              }, 1500);
            }
          }
        });
        FileObserver watch = watchModifier.get(newFile.getPath());
        if(watch!=null) watch.startWatching();
      }
    }
  }

  void setScreenRecordStatus(boolean value){
    if(screenRecording != value){
      new Handler(Looper.getMainLooper()).post(() -> {
        screenRecording = value;
        channel.invokeMethod("screenrecord",value);
      });
    }
  }

  @Override
  public void onDetachedFromEngine(@NonNull FlutterPluginBinding binding) {
  }


  public static String getMimeType(String url) {
    String type = null;
    String extension = MimeTypeMap.getFileExtensionFromUrl(url);
    if (extension != null) {
      type = MimeTypeMap.getSingleton().getMimeTypeFromExtension(extension);
    }
    return type;
  }

  public static File getLastModified(String directoryFilePath)
  {
    File directory = new File(directoryFilePath);
    if(directory.listFiles() == null) return null;
    File[] files = directory.listFiles(File::isFile);
    long lastModifiedTime = Long.MIN_VALUE;
    File chosenFile = null;

    if (files != null)
    {
      for (File file : files)
      {
        if (file.lastModified() > lastModifiedTime)
        {
          chosenFile = file;
          lastModifiedTime = file.lastModified();
        }
      }
    }

    return chosenFile;
  }

  @Override
  public void onAttachedToActivity(@NonNull ActivityPluginBinding binding) {
    activityPluginBinding = binding;

  }

  @Override
  public void onDetachedFromActivityForConfigChanges() {

  }

  @Override
  public void onReattachedToActivityForConfigChanges(@NonNull ActivityPluginBinding binding) {

  }

  @Override
  public void onDetachedFromActivity() {

  }

  public enum Path {
    DCIM(Environment.getExternalStoragePublicDirectory(Environment.DIRECTORY_DCIM) + File.separator + "Screenshots" + File.separator),
    PICTURES(Environment.getExternalStoragePublicDirectory(Environment.DIRECTORY_PICTURES ) + File.separator + "Screenshots" + File.separator);

    final private String path;

    public String getPath() {
      return path;
    }

    Path(String path) {
      this. path = path;
    }
  }

}

File: packages/disable_screenshot/android/src/main/kotlin/com/appinio/disable_screenshot/DisableScreenshotPlugin.kt
package com.appinio.disable_screenshot

import android.app.Activity
import android.content.Context
import android.util.Log
import android.view.WindowManager
import androidx.annotation.NonNull;

import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import io.flutter.plugin.common.PluginRegistry.Registrar
import kotlin.coroutines.coroutineContext

/** DisableScreenshotsPlugin */
public class DisableScreenshotPlugin: FlutterPlugin, MethodCallHandler, EventChannel.StreamHandler, ActivityAware {
  /// The MethodChannel that will the communication between Flutter and native Android
  ///
  /// This local reference serves to register the plugin with the Flutter Engine and unregister it
  /// when the Flutter Engine is detached from the Activity
  private lateinit var channel : MethodChannel
  private lateinit var applicationContext: Context;
  private lateinit var activity: Activity;
  private var eventSink: EventChannel.EventSink? = null
  private lateinit var screenShotListenManager: ScreenShotListenManager;
  var disableScreenshots: Boolean = false

  override fun onAttachedToEngine(@NonNull flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {

    onAttachedToEngine(flutterPluginBinding.applicationContext, flutterPluginBinding.binaryMessenger)
  }

  private fun onAttachedToEngine(applicationContext: Context, messenger: BinaryMessenger) {
    this.applicationContext = applicationContext
    this.channel = MethodChannel(messenger, "com.appinio.screenshot/disableScreenshots")
    val eventChannel = EventChannel(messenger, "com.appinio.screenshot/observer")
    this.channel.setMethodCallHandler(this)
    eventChannel.setStreamHandler(this)
  }

  private fun setDisableScreenshotsStatus(disable: Boolean) {
    if (disable) {
      activity.window.setFlags(WindowManager.LayoutParams.FLAG_SECURE, WindowManager.LayoutParams.FLAG_SECURE);
      println("disable screenshot")
    } else {
      activity.window.clearFlags(WindowManager.LayoutParams.FLAG_SECURE)
      println("allow screenshot")
    }
  }

  override fun onMethodCall(@NonNull call: MethodCall, @NonNull result: Result) {
    if (call.method == "disableScreenshots") {
      var disable = call.argument<Boolean>("disable") == true
      setDisableScreenshotsStatus(disable)
      result.success("")
    } else {
      result.notImplemented()
    }
  }

  override fun onDetachedFromEngine(@NonNull binding: FlutterPlugin.FlutterPluginBinding) {
    channel.setMethodCallHandler(null)
  }

  override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
    println("start listening")
    eventSink = events
    screenShotListenManager = ScreenShotListenManager.newInstance(applicationContext)
    screenShotListenManager.setListener { imagePath ->
      println("got screenshot with path：$imagePath")
      eventSink?.success("got screenshot")
    }
    screenShotListenManager.startListen()
  }

  override fun onCancel(arguments: Any?) {
    screenShotListenManager.stopListen()
    eventSink = null
  }


  override fun onDetachedFromActivity() {

  }

  override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {
    activity = binding.activity
    setDisableScreenshotsStatus(this.disableScreenshots)
  }

  override fun onAttachedToActivity(binding: ActivityPluginBinding) {
    activity = binding.activity
    setDisableScreenshotsStatus(this.disableScreenshots)
  }

  override fun onDetachedFromActivityForConfigChanges() {

  }
}
